Grpc.Tools causes a build error when .proto file name contains dot

<!--

This form is for bug reports and feature requests ONLY!
For general questions and troubleshooting, please ask/look for answers here:
- grpc.io mailing list: https://groups.google.com/forum/#!forum/grpc-io
- StackOverflow, with "grpc" tag: http://stackoverflow.com/questions/tagged/grpc

Issues specific to *grpc-java*, *grpc-go*, *grpc-node*, *grpc-dart*, *grpc-web* should be created in the repository they belong to (e.g. https://github.com/grpc/grpc-LANGUAGE/issues/new)
-->
 
### What version of gRPC and what language are you using?
 
Grpc 1.17.1
Grpc.Core 1.17.1
Grpc.Tools 1.17.1

Language: C#
 
### What operating system (Linux, Windows, …) and version?
 
Windows 10 Pro 1803 64bit
 
### What runtime / compiler are you using (e.g. python version or version of gcc)
 
.NET Core SDK 2.1.502
 
### What did you do?

I tried to build .proto  which file name contains dot(e.g. hello.world.proto) by Grpc.Tools.

source files: [Greeter.zip](https://github.com/grpc/grpc/files/2737695/Greeter.zip)
 
### What did you expect to see?
 
Build succeeds.
 
### What did you see instead?

Following build errors:
```
obj\Debug\netstandard1.5\Hello.worldGrpc.cs(18,57,18,69): error CS0234: The type or namespace name 'HelloRequest' does not exist in the namespace 'Helloworld' (are you missing an assembly reference?)
obj\Debug\netstandard1.5\Hello.worldGrpc.cs(19,57,19,67): error CS0234: The type or namespace name 'HelloReply' does not exist in the namespace 'Helloworld' (are you missing an assembly reference?)
obj\Debug\netstandard1.5\Hello.worldGrpc.cs(21,53,21,65): error CS0234: The type or namespace name 'HelloRequest' does not exist in the namespace 'Helloworld' (are you missing an assembly reference?)
obj\Debug\netstandard1.5\Hello.worldGrpc.cs(21,86,21,96): error CS0234: The type or namespace name 'HelloReply' does not exist in the namespace 'Helloworld' (are you missing an assembly reference?)
obj\Debug\netstandard1.5\Hello.worldGrpc.cs(43,117,43,129): error CS0234: The type or namespace name 'HelloRequest' does not exist in the namespace 'Helloworld' (are you missing an assembly reference?)
obj\Debug\netstandard1.5\Hello.worldGrpc.cs(43,77,43,87): error CS0234: The type or namespace name 'HelloReply' does not exist in the namespace 'Helloworld' (are you missing an assembly reference?)
obj\Debug\netstandard1.5\Hello.worldGrpc.cs(81,80,81,92): error CS0234: The type or namespace name 'HelloRequest' does not exist in the namespace 'Helloworld' (are you missing an assembly reference?)
obj\Debug\netstandard1.5\Hello.worldGrpc.cs(81,41,81,51): error CS0234: The type or namespace name 'HelloReply' does not exist in the namespace 'Helloworld' (are you missing an assembly reference?)
obj\Debug\netstandard1.5\Hello.worldGrpc.cs(91,80,91,92): error CS0234: The type or namespace name 'HelloRequest' does not exist in the namespace 'Helloworld' (are you missing an assembly reference?)
obj\Debug\netstandard1.5\Hello.worldGrpc.cs(91,41,91,51): error CS0234: The type or namespace name 'HelloReply' does not exist in the namespace 'Helloworld' (are you missing an assembly reference?)
obj\Debug\netstandard1.5\Hello.worldGrpc.cs(103,107,103,119): error CS0234: The type or namespace name 'HelloRequest' does not exist in the namespace 'Helloworld' (are you missing an assembly reference?)
obj\Debug\netstandard1.5\Hello.worldGrpc.cs(103,62,103,72): error CS0234: The type or namespace name 'HelloReply' does not exist in the namespace 'Helloworld' (are you missing an assembly reference?)
obj\Debug\netstandard1.5\Hello.worldGrpc.cs(113,107,113,119): error CS0234: The type or namespace name 'HelloRequest' does not exist in the namespace 'Helloworld' (are you missing an assembly reference?)
```

--- a/src/csharp/Grpc.Tools.Tests/CSharpGeneratorTest.cs
+++ b/src/csharp/Grpc.Tools.Tests/CSharpGeneratorTest.cs
@@ -33,12 +33,14 @@ namespace Grpc.Tools.Tests
         [TestCase("foo.proto", "Foo.cs", "FooGrpc.cs")]
         [TestCase("sub/foo.proto", "Foo.cs", "FooGrpc.cs")]
         [TestCase("one_two.proto", "OneTwo.cs", "OneTwoGrpc.cs")]
-        [TestCase("__one_two!.proto", "OneTwo!.cs", "OneTwo!Grpc.cs")]
-        [TestCase("one(two).proto", "One(two).cs", "One(two)Grpc.cs")]
-        [TestCase("one_(two).proto", "One(two).cs", "One(two)Grpc.cs")]
-        [TestCase("one two.proto", "One two.cs", "One twoGrpc.cs")]
-        [TestCase("one_ two.proto", "One two.cs", "One twoGrpc.cs")]
-        [TestCase("one .proto", "One .cs", "One Grpc.cs")]
+        [TestCase("ONE_TWO.proto", "ONETWO.cs", "ONETWOGrpc.cs")]
+        [TestCase("one.two.proto", "OneTwo.cs", "One.twoGrpc.cs")]
+        [TestCase("__one_two!.proto", "OneTwo.cs", "OneTwo!Grpc.cs")]
+        [TestCase("one(two).proto", "OneTwo.cs", "One(two)Grpc.cs")]
+        [TestCase("one_(two).proto", "OneTwo.cs", "One(two)Grpc.cs")]
+        [TestCase("one two.proto", "OneTwo.cs", "One twoGrpc.cs")]
+        [TestCase("one_ two.proto", "OneTwo.cs", "One twoGrpc.cs")]
+        [TestCase("one .proto", "One.cs", "One Grpc.cs")]
         public void NameMangling(string proto, string expectCs, string expectGrpcCs)
         {
             var poss = _generator.GetPossibleOutputs(Utils.MakeItem(proto, "grpcservices", "both"));

--- a/src/csharp/Grpc.Tools.Tests/CSharpGeneratorTest.cs
+++ b/src/csharp/Grpc.Tools.Tests/CSharpGeneratorTest.cs
@@ -35,6 +35,7 @@ namespace Grpc.Tools.Tests
         [TestCase("one_two.proto", "OneTwo.cs", "OneTwoGrpc.cs")]
         [TestCase("ONE_TWO.proto", "ONETWO.cs", "ONETWOGrpc.cs")]
         [TestCase("one.two.proto", "OneTwo.cs", "One.twoGrpc.cs")]
+        [TestCase("one123two.proto", "One123Two.cs", "One123twoGrpc.cs")]
         [TestCase("__one_two!.proto", "OneTwo.cs", "OneTwo!Grpc.cs")]
         [TestCase("one(two).proto", "OneTwo.cs", "One(two)Grpc.cs")]
         [TestCase("one_(two).proto", "OneTwo.cs", "One(two)Grpc.cs")]

--- a/src/csharp/Grpc.Tools/GeneratorServices.cs
+++ b/src/csharp/Grpc.Tools/GeneratorServices.cs
@@ -66,29 +66,28 @@ namespace Grpc.Tools
         public override string[] GetPossibleOutputs(ITaskItem protoItem)
         {
             bool doGrpc = GrpcOutputPossible(protoItem);
-            string filename = LowerUnderscoreToUpperCamel(
-                Path.GetFileNameWithoutExtension(protoItem.ItemSpec));
-
             var outputs = new string[doGrpc ? 2 : 1];
+            string basename = Path.GetFileNameWithoutExtension(protoItem.ItemSpec);
+
             string outdir = protoItem.GetMetadata(Metadata.OutputDir);
-            string fileStem = Path.Combine(outdir, filename);
-            outputs[0] = fileStem + ".cs";
+            string filename = LowerUnderscoreToUpperCamelProtocWay(basename);
+            outputs[0] = Path.Combine(outdir, filename) + ".cs";
+
             if (doGrpc)
             {
                 // Override outdir if kGrpcOutputDir present, default to proto output.
-                outdir = protoItem.GetMetadata(Metadata.GrpcOutputDir);
-                if (outdir != "")
-                {
-                    fileStem = Path.Combine(outdir, filename);
-                }
-                outputs[1] = fileStem + "Grpc.cs";
+                string grpcdir = protoItem.GetMetadata(Metadata.GrpcOutputDir);
+                filename = LowerUnderscoreToUpperCamelGrpcWay(basename);
+                outputs[1] = Path.Combine(
+                    grpcdir != "" ? grpcdir : outdir, filename) + "Grpc.cs";
             }
             return outputs;
         }
 
-        string LowerUnderscoreToUpperCamel(string str)
+        // This is how the gRPC codegen currently construct its output filename.
+        // See src/compiler/generator_helpers.h:118.
+        string LowerUnderscoreToUpperCamelGrpcWay(string str)
         {
-            // See src/compiler/generator_helpers.h:118
             var result = new StringBuilder(str.Length, str.Length);
             bool cap = true;
             foreach (char c in str)
@@ -106,6 +105,26 @@ namespace Grpc.Tools
                 {
                     result.Append(c);
                 }
+            }
+            return result.ToString();
+        }
+
+        // This is how the protoc codegen constructs its output filename.
+        // See protobuf/compiler/csharp/csharp_helpers.cc:356.
+        // Note that protoc explicitly discards non-ASCII letters.
+        string LowerUnderscoreToUpperCamelProtocWay(string str)
+        {
+            var result = new StringBuilder(str.Length, str.Length);
+            bool cap = true;
+            foreach (char c in str)
+            {
+                char upperC = char.ToUpperInvariant(c);
+                bool isAsciiLetter = 'A' <= upperC && upperC <= 'Z';
+                if (isAsciiLetter || ('0' <= c && c <= '9'))
+                {
+                    result.Append(cap ? upperC : c);
+                }
+                cap = !isAsciiLetter;
             }
             return result.ToString();
         }

--- a/src/csharp/Grpc.Tools/GeneratorServices.cs
+++ b/src/csharp/Grpc.Tools/GeneratorServices.cs
@@ -110,7 +110,7 @@ namespace Grpc.Tools
         }
 
         // This is how the protoc codegen constructs its output filename.
-        // See protobuf/compiler/csharp/csharp_helpers.cc:356.
+        // See protobuf/compiler/csharp/csharp_helpers.cc:137.
         // Note that protoc explicitly discards non-ASCII letters.
         string LowerUnderscoreToUpperCamelProtocWay(string str)
         {

